packages/math-poolsimulator/src/index.ts

```diff
@@ -26,266 +27,166 @@ export function numberToBigNumber({
-          amm.poolParams as IMetaStableMath
-        );
+export class AMM<TPoolPairData extends PoolPairData> {
+  private math: IAMMFunctionality<TPoolPairData>;
 
-        break;
-      case "GyroE":
-        this.math = new ExtendedGyroEV2(amm.poolParams as IGyroEMaths);
-        break;
+  constructor(amm: IAMM<TPoolPairData>) {
+    const factory = mathFactories[amm.poolType as PoolType];
+    if (!factory) {
+        throw new Error(`Unsupported poolType: ${amm.poolType}`);
     }
-  }
+    this.math = factory(amm.poolParams);
+}
 
-  exactTokenInForTokenOut(
-    amountIn: number,
-    tokenIn: string,
-    tokenOut: string
-  ): number {
-    if (this.math instanceof ExtendedMetaStableMath) {
-      const poolPairData = this.math.parsePoolPairData(
-        tokenIn,
-        tokenOut
-      ) as MetaStablePoolPairData;
-      return this.math
-        ._exactTokenInForTokenOut(poolPairData, bnum(amountIn))
-        .toNumber();
-    } else if (this.math instanceof ExtendedGyroEV2) {
-      const poolPairData = this.math.parsePoolPairData(
-        tokenIn,
-        tokenOut
-      ) as GyroEPoolPairData;
-      return this.math
-        ._exactTokenInForTokenOut(poolPairData, bnum(amountIn))
-        .toNumber();
-    } else throw new Error("Invalid math instance");
+  exactTokenInForTokenOut(amountIn: number, tokenIn: string, tokenOut: string): number {
+    const poolPairData = this.math.parsePoolPairData(tokenIn, tokenOut);
+    return this.math._exactTokenInForTokenOut(poolPairData, amountIn);
   }
 
   tokenInForExactTokenOut(
     amountOut: number,
     tokenIn: string,
     tokenOut: string
   ): number {
-    if (this.math instanceof ExtendedMetaStableMath) {
       const poolPairData = this.math.parsePoolPairData(
         tokenIn,
         tokenOut
-      ) as MetaStablePoolPairData;
-
-      return this.math
-        ._tokenInForExactTokenOut(poolPairData, bnum(amountOut))
-        .toNumber();
-    } else if (this.math instanceof ExtendedGyroEV2) {
-      const poolPairData = this.math.parsePoolPairData(
-        tokenIn,
-        tokenOut
-      ) as GyroEPoolPairData;
+      );
 
       return this.math
         ._tokenInForExactTokenOut(poolPairData, bnum(amountOut))
         .toNumber();
-    } else throw new Error("Invalid math instance");
   }
 
   spotPrice(tokenIn: string, tokenOut: string): number {
-    if (this.math instanceof ExtendedMetaStableMath) {
       const poolPairData = this.math.parsePoolPairData(
         tokenIn,
         tokenOut
-      ) as MetaStablePoolPairData;
-      return this.math._spotPrice(poolPairData).toNumber();
-    } else if (this.math instanceof ExtendedGyroEV2) {
-      const poolPairData = this.math.parsePoolPairData(
-        tokenIn,
-        tokenOut
-      ) as GyroEPoolPairData;
+      );
       return this.math._spotPrice(poolPairData).toNumber();
-    } else throw new Error("Invalid math instance");
   }
 
-  _tokenInForExactSpotPriceAfterSwap({
```

Comment:
This one is commented out because it really changes depending on the pool type right?